Source/DafnyStandardLibraries/src/DafnyStdLibs/DynamicArray.dfy

```diff
@@ -0,0 +1,218 @@
+  */
+  class DynamicArray<A> {
+    ghost var items: seq<A>
+    ghost var Repr: set<object>
+
+    const defaultValue: A
+    var size: nat
+    var capacity: nat
+    var data: array<A>
+
+    ghost predicate Valid?()
+      reads this, Repr
+    {
+      && Repr == {this, data}
+      && capacity != 0
+      && data.Length == capacity as int
+      && size <= capacity
+      && size as int == |items|
+      && items == data[..size]
+    }
+
+    constructor(defaultValue: A, initialCapacity: nat := 8)
+      requires initialCapacity > 0
+      ensures size == 0
+      ensures items == []
+      ensures fresh(Repr)
+      ensures Valid?()
+      ensures capacity == initialCapacity
+    {
+      this.defaultValue := defaultValue;
+      items := [];
+      size := 0;
+      capacity := initialCapacity;
+      data := new A[initialCapacity](_ => defaultValue);
+      Repr := {this, data};
+    }
+
+    /**
+    Retrieve the element at index 
+     */
+    function At(index: nat) : (element: A)
+      reads this, Repr
+      requires index < size
+      requires Valid?()
+      ensures element == items[index]
+    {
+      data[index]
+    }
+
+    /**
+    Put element in the array at index 
+     */
+    method Put(index: nat, element: A)
+      requires index < size
+      requires Valid?()
+      modifies Repr, `items
+      ensures Valid?()
+      ensures fresh(Repr - old(Repr))
+      ensures size == old(size)
+      ensures items == old(items)[index := element]
+    {
+      data[index] := element;
+      items := items[index := element];
+    }
+
+    /**
+    Ensure that at least a reserved amount of elements can still be pushed onto the array in constant time
+
+    Returns false only if it was not at all possible to provide the ensurance.
+     */
+    method Ensure(reserved: nat)
+      requires Valid?()
+      ensures Valid?()
+      modifies `capacity, Repr
+      ensures size == old(size)
+      ensures items == old(items)
+      ensures fresh(Repr - old(Repr))
+      ensures reserved <= capacity - size
+    {
+      var newCapacity := capacity;
+      while reserved > newCapacity - size
+        invariant newCapacity >= capacity
+      {
+        newCapacity := DefaultNewCapacity(newCapacity);
+      }
+      if (newCapacity > capacity) {
+        Realloc(newCapacity);
+      }
+    }
+
+    /**
+    Pop an element from the array in constant time
+     */
+    method PopFast()
+      requires Valid?()
+      requires size > 0
+      modifies `size, `items
+      ensures Valid?()
+      ensures size < capacity
+      ensures size == old(size) - 1
+      ensures capacity == old(capacity)
+      ensures items == old(items[..|items| - 1])
+    {
+      size := size - 1;
+      items := items[..|items| - 1];
+    }
+
+    /**
+    Push an element onto the array in constant time
+     */
+    method PushFast(element: A)
+      requires Valid?()
+      requires size < capacity
+      modifies Repr, `size, `items
```

Comment:
Since `this` is already part of `Repr`, you don't need to list `` `size, `items``

Reply:
Same for method `Ensure`.